Quartz/Engine/Include/Quartz/Utilities/Commander.hpp

```diff
@@ -0,0 +1,116 @@
+// Copyright 2019 Genten Studios
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// 1. Redistributions of source code must retain the above copyright notice,
+// this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#pragma once
+
+#include <string>
+#include <array>
+#include <functional>
+
+namespace qz
+{
+	namespace utils
+	{
+        constexpr int MaxCommandsNumber = 100;
+        constexpr int MaxArgumentNumber = 10;
+
+        /**
+         * @brief The command book stores commands and information on them to be used by a commander
+         */
+
+        struct CommandBook{
+            std::array<std::string, MaxCommandsNumber> m_command;
+            std::array<std::string, MaxCommandsNumber> m_help;
+            std::array<std::string, MaxCommandsNumber> m_permission;
+            std::array<std::function<int(std::array<std::string, MaxArgumentNumber> args)>, MaxCommandsNumber> m_functions;
```

Comment:
It may make this cleaner to typedef the `std::function<...>` type out.
Something like this
```cpp
typedef std::function<int(std::array<std::string, MaxArgumentNumber>)> CommandAction;
...
std::array<CommandAction, MaxCommandsNumber> m_functions;
```

Reply:
`using` is more c++-ish

Reply:
Yeah, I did wonder that, but just defaulted back to the C way :D

Something more like:
```cpp
using CommandAction = std::function<int(std::array<std::string, MaxArgumentNumber>)>;
```
(or something like that, I cannot remember the syntax)